---
title: "_Stb6_ and _Stb15_ haplotype heatmaps"
author: "Amber N. Hafeez"
output: html_document
---

This notebook details the process of using a SNP distance matrix of a gene/locus and phenotype data to generate a heatmap using `pheatmap` and subsequently identify alleles of interest.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Data preparation for `pheatmap`

Load file with reference to different line names. This file only contains the lines of interest, so will be used for filtering later.
```{r}
wnames <- read.csv("/Users/hafeeza/Documents/WatSeq/Watcodes.csv", strip.white = T)
head(wnames)
```

Load the package `pheatmap`:
```{r}
if(!require(pheatmap)){install.packages("pheatmap")}
library(pheatmap)
```

Load the distance matrix for _Stb6_ generated by running the `vcf_parse_distance.py` script. Set rownames as the accession name.
```{r}
mx <- read.csv("/Users/hafeeza/Documents/WatSeq/vcf_parse/W300_Stb6_distmatrix_wholegene.tsv", sep="\t")

mx2 <- mx[,-1]
row.names(mx2) <- mx$Accessions
```

Filter out any lines which are not in the `wnames` dataframe. 
```{r}
fmx <- mx2[row.names(mx2) %in% wnames$WATDE, colnames(mx2) %in% wnames$WATDE]
```

Read the pathology data in (estimated mean AUDPC of pycnidia (logit) and damage (% of maximum AUDPC)). Then combine this with the `wnames` dataframe:
```{r}
pt<-read.csv('/Users/hafeeza/Documents/Septoria experiments/Watkins/Core_300_comb/Tables of estimates/IPO323_IPO88004_IPO90012_est_means_pmaxdAUDPC.csv',head=T)
pt<-pt[,-1]
colnames(wnames)[1] <- 'Line'
pt<-merge(pt,wnames,all.x=T)
```

### Pheatmap analysis for _Stb6_

Filter for IPO323 pathology results and generate a dataframe (`pto_2`) which just contains the `pAUDPC` and `%max_dAUDPC` values. This will be used to annotate the `pheatmap` figure.
```{r}
pt3 <- pt[pt$Isolate=='IPO323',]
pto <- pt3[match(colnames(fmx),pt3$WATDE),]
row.names(pto) <- pto$WATDE
pto_3 <- pto[,3:4]
colnames(pto_3) <- c('pAUDPC','%max_dAUDPC')
```

Generate the `pheatmap` figure using the SNP distance matrix (`fmx`) and annotations (`pto_3`). 
```{r}
phmStb6 <- pheatmap(fmx,
         annotation_row = pto_3,
         fontsize_row=3.5, fontsize_col=3.5)
```
From the pheatmap object, extract the clusters as well as the distance from key reference lines Chinese Spring and Longbow:
```{r}
stb6c <- cbind(cbind(rownames(fmx),fmx$Chinese_Spring,fmx$Longbow), 
                      cluster = cutree(phmStb6$tree_row, k=10))

stb6c <- data.frame(stb6c)
colnames(stb6c) <- c("WATDE","dist_CS","dist_Longbow","cluster")

table(stb6c$cluster)

stb6c <- merge(stb6c, wnames[,1:2])

pto_3$WATDE <- row.names(pto_3)
pt_stb6 <- merge(stb6c, pto_3)
colnames(pt_stb6)[7] <- "pmax_dAUDPC"

pt_stb6$cluster <-factor(pt_stb6$cluster, levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10"))
```

```{r}
if(!require(ggplot2)){install.packages(ggplot2)}
library(ggplot2)
if(!require(cowplot)){install.packages(cowplot)}
library(cowplot)
if(!require(ggrepel)){install.packages(ggrepel)}
library(ggrepel)
```

```{r}
ggplot(pt_stb6, aes(x= pmax_dAUDPC, y=pAUDPC, color=cluster)) +
  geom_point()+
  theme_cowplot(12) + background_grid(major="xy", minor="none") +
  scale_color_brewer(palette = "Paired") +
  geom_label_repel(data = pt_stb6[pt_stb6$Line %in% c("Chinese Spring", "ArinaLrFor","Cadenza","Norin 61","CDC Landmark","CDC Stanley","SY Mattis","Lancer","Mace","Jagger","Weebill","Julius", "Claire", "Robigus", "Paragon","Longbow","Flame", "Baj"),], aes(label=Line), 
                   point.padding = 0.2, size=2.5, box.padding=0.3,segment.size = 0.3, show.legend = FALSE) +
  xlab("% max. dAUDPC") + ylab("Logit pAUDPC") + labs(color="Cluster Name") +
  guides(color = guide_legend(override.aes = list(shape = 19)))
```
It is known that the Chinese Spring and Cadenza alleles of _Stb6_ are functional and that the Longbow allele is non-functional/confer susceptibility. The above graph demonstrates that this is broadly true in lines tested, however there is some variation in resistance of lines carrying resistant alleles; groups 1 and 2 contain 13 susceptible lines. The function of the other eight groups is not clear as lines with these alleles vary in their resistance to IPO323, or only contain a single accession (groups 9 and 10).

In conclusion, groups 1 and 2 can be considered to have functional alleles of _Stb6_.

### Repeat the same steps to conduct the pheatmap analysis for TraesCS6A02G078700
```{r}
isolate <- 'IPO88004'
matrix <- read.csv("/Users/hafeeza/Documents/WatSeq/vcf_parse/stb15/TraesCS6A02G078700_distmatrix_wholegene.tsv", sep="\t")
hm <- matrix[,-1]
row.names(hm) <- matrix$Accessions

pt <- pt[pt$Isolate==isolate,]
fhm <- hm[row.names(hm) %in% wnames$WATDE, colnames(hm) %in% wnames$WATDE]
pto <- pt[match(colnames(fhm),pt$WATDE),]
row.names(pto) <- pto$WATDE
pto2 <- pto[,3:4]
colnames(pto2) <- c('pAUDPC','%max_dAUDPC')

stb15phm <- pheatmap(fhm,
         annotation_row = pto2,
         fontsize_row=3.5, fontsize_col=3.5)

stb15phm
```
From the pheatmap object, extract the clusters as well as the distance from key reference lines Chinese Spring and ArinaLrFor:
```{r}
# list of accessions with their cluster number
# can change k to make clusters more or less defined
stb15.clust <- cbind(cbind(rownames(fhm),fhm$Chinese_Spring,fhm$ArinaLrFor), 
                      cluster = cutree(stb15phm$tree_row, k=10))

stb15.clust <- data.frame(stb15.clust)
colnames(stb15.clust) <- c("WATDE","dist_CS","dist_Arina","cluster")

pt2 <- merge(pt, stb15.clust, by='WATDE', all.x=T)
pt_stb15 <- pt2[pt2$Isolate=="IPO88004",]
head(pt_stb15)
```
The clusters were given meaninful names based on whether they contained any wheat cultivars:
```{r}
pt_stb15$clust_name <- NA
pt_stb15$clust_name <- as.character(pt_stb15$clust_name)

pt_stb15[pt_stb15$cluster %in% "1",]$clust_name <- "ArinaLrFor"
pt_stb15[pt_stb15$cluster %in% "2",]$clust_name <- "Chinese Spring"
pt_stb15[pt_stb15$cluster %in% "3",]$clust_name <- "Robigus"
pt_stb15[pt_stb15$cluster %in% "4",]$clust_name <- "WatUnique 1"
pt_stb15[pt_stb15$cluster %in% "5",]$clust_name <- "WatUnique 2"
pt_stb15[pt_stb15$cluster %in% "6",]$clust_name <- "WatUnique 3"
pt_stb15[pt_stb15$cluster %in% "7",]$clust_name <- "WatUnique 4"
pt_stb15[pt_stb15$cluster %in% "8",]$clust_name <- "WatUnique 5"
pt_stb15[pt_stb15$cluster %in% "9",]$clust_name <- "WatUnique 6"
pt_stb15[pt_stb15$cluster %in% "10",]$clust_name <- "Claire"

pt_stb15$Type <- ifelse(grepl("^W", pt_stb15$Line), "Watkins Landrace", "Cultivar")
```

The VCFs from which SNP distance matrices were generated were studied in IGV. This revealed that some groups only different from eachother in positions of missing data, and were determined to not be different from eachother in a meaningful way. WatUnique 3 was not included in any group as it consisted of a single line with a lot of missing data.

They were therefore combined to form four groups as below. This lesser number of groups also seemed a better fit given study of the heatmap which has four main groups, two of them (the Arina-Robigus and Chinese Spring groups) representing the vast majority of lines.

```{r}
pt_stb15[pt_stb15$clust_name %in% c("ArinaLrFor", "Robigus", "Claire"),]$clust_name <- "Arina-Robigus"
pt_stb15[pt_stb15$clust_name %in% c("WatUnique 1", "WatUnique 4"),]$clust_name <- "Chinese Spring"
pt_stb15[pt_stb15$clust_name %in% "WatUnique 3",]$clust_name <- NA
pt_stb15[pt_stb15$Line %in% "W460",]$clust_name <- "Arina-Robigus"

pt_stb15[pt_stb15$clust_name %in% "WatUnique 2",]$clust_name <- "WatUnique 1"
pt_stb15[pt_stb15$clust_name %in% c("WatUnique 5", "WatUnique 6"),]$clust_name <- "WatUnique 2"

table(pt_stb15$clust_name)
```

Plotting a damage vs pycnidia plot for IPO88004 with colours of points corresponding to their _Stb15_ allele.

[colour palettes](https://www.datanovia.com/en/blog/top-r-color-palettes-to-know-for-great-data-visualization/)
```{r}
library(RColorBrewer)
library(ggrepel)
ggplot(pt_stb15, aes(x=emmean_pmax_dAUDPC, y=emmean_lgt_pAUDPC, color=clust_name)) +
  geom_point()+
  theme_cowplot(12) + background_grid(major="xy", minor="none") +
  scale_color_brewer(palette = "PiYG") +
  geom_label_repel(data = pt_stb15[pt_stb15$Line %in% c("Chinese Spring", "ArinaLrFor","Cadenza","Norin 61","CDC Landmark","CDC Stanley","SY Mattis","Lancer","Mace","Jagger","Weebill","Julius", "Claire", "Robigus", "Paragon"),], aes(label=Line), 
                   point.padding = 0.2, size=2.5, box.padding=0.3,segment.size = 0.3, show.legend = FALSE) +
  xlab("% max. dAUDPC") + ylab("Logit pAUDPC") + labs(color="Cluster Name")

```

All lines in the Arina-Robigus group are resistant. This was good evidence that this allele of the candidate gene TraesCS6A02G078700 corresponded to _Stb15_. The Chinese Spring group is broadly susceptible as expected, but does contain some resistant accessions. This could be explained by other resistances to IPO88004. 


